current/runtime/storage.md

```diff
@@ -4,22 +4,128 @@ lang: en
+[The `storage` module in the Substrate Support pallet](https://substrate.dev/rustdocs/master/frame_support/storage/index.html)
+gives runtime developers access to Substrate's flexible storage APIs. Any value which can be encoded by the
+[Parity SCALE codec](../advanced/codec) is supported by these storage APIs:
 
-* [Storage Value](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html) - A single value.
-* [Storage Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html) - A key-value hash map.
-* [Storage Linked Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageLinkedMap.html) - Similar to a storage map, but allows enumeration of the stored elements.
-* [Storage Double Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html) - An implementation of a map with two keys.
+* [Storage Value](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html) - A single value
+* [Storage Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html) - A key-value hash map
+* [Storage Double Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html) - An
+  implementation of a map with two keys that provides the important ability to efficiently remove all entries that have
+  a common first key
+* Iterable Storage Maps - Implementations of
+  [Storage Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageMap.html) and
+  [Storage Double Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageDoubleMap.html)
+  whose keys and values can be iterated over
 
-Any value which can be encoded by the [Parity SCALE codec](../advanced/codec) is supported by these storage APIs.
+The type of Storage Item you select should depend on the logical way in which the value will be used by your runtime.
 
-### Storage Declaration
+### Storage Value
 
-You can use the `decl_storage!` macro to easily create new runtime storage items. Here is an example of what it looks like to declare each type of storage item:
+This type of storage item should be used for values that are viewed as a single unit by the runtime, whether that is a
+single primitive value, a single `struct` or a single collection of related items. Although wrapping related items in
+a shared `struct` is an excellent way to reduce the number of storage reads (an important consideration), at some point
+the size of the object will begin to incur costs that may outweigh the optimization in storage reads. Storage values
+can be used to store lists of items, but runtime developers should take care with respect to the size of these lists.
+Large lists incur storage costs just like large `structs`. Furthermore, iterating over a large list in your runtime may
+result in exceeding the block production time.
```

Comment:
I believe that Shawn has indicated we should specifically caution people against iteration within the runtime.

Reply:
(ok but just to clarify if you store in your list something different from basic u32, u64... then decoding/encoding is done by iterating other elements.)

Reply:
Can you please expand on that concept just a bit for me? Are you referring to the fact that anything other than a Rust primitive requires a pointer to be dereferenced?

Reply:
in parity-scale-codec to decode `Vec<u32>` and all other type in `	U8, I8, U16, I16, U32, I32, U64, I64, U128, I128,` we decode/encode by doing mem::transmute, otherwise we do:
```rust
fn encode(..) {
...
				for item in self {
					item.encode_to(dest);
				}
}
```
and 
```rust
fn decode(...) {
...
					for _ in 0..len {
						r.push(T::decode(input)?);
					}
}
```

Reply:
I see, so complex data types require an extra layer of indirection that materializes as an encoding/decoding step. Correct?

Reply:
I wouldn't phrases this way. Just encoding/decoding of vec is optimized for `U8, I8, U16, I16, U32, I32, U64, I64, U128, I128` I think also Option is optimized for them and some other generic data type.
But this is just optimization in my point of view.

Reply:
Do you think that it is important for us to document this optimization with respect to the runtime storage interface?

Reply:
not sure, should belong to some codec doc I think

Reply:
Opened a separate issue :+1: #43